### ISSUE TITLE
AWS 2-5: Style TDX Streams Appropriately

### ISSUE THREAD

**Issue by rajadain** [WikiWatershed/model-my-watershed#3625]
Currently, the TDX streams look very dense when zoomed out:

<img width="1537" alt="image" src="https://github.com/WikiWatershed/model-my-watershed/assets/1430060/f5b6e64c-4e6c-4b33-a3e2-fa620ccce71e">

Compare this with US Medium Resolution Streams, of which only the higher order streams are shown at low zoom levels:

<img width="1537" alt="image" src="https://github.com/WikiWatershed/model-my-watershed/assets/1430060/407dadc5-a9a8-436d-bed0-2b692704de59">

This communicates the relationships and importance of streams very clearly.

This information is in the dataset:

```
ogrinfo TDX_streamnet_7020038340_01.gpkg TDX_streamnet_7020038340_01

OGRFeature(TDX_streamnet_7020038340_01):10744
  LINKNO (Integer) = 49200
  DSLINKNO (Integer) = 106032
  USLINKNO1 (Integer) = 48608
  USLINKNO2 (Integer) = 48016
  DSNODEID (Integer64) = -1
  strmOrder (Integer) = 2
  Length (Real) = 641.6
  Magnitude (Integer) = 2
  DSContArea (Real) = 10775170
  strmDrop (Real) = 6.08
  Slope (Real) = 0.00947399913
  StraightL (Real) = 491.4
  USContArea (Real) = 10652958
  WSNO (Integer) = 49200
  DOUTEND (Real) = 95675
  DOUTSTART (Real) = 96316.5
  DOUTMID (Real) = 95995.8
  LINESTRING (-70.289999999999 44.0646666666665,-70.2901111111101 44.0646666666665,-70.2902222222212 44.0646666666665,-70.2903333333323 44.0645555555554,-70.2904444444435 44.0644444444443,-70.2905555555546 44.0643333333332,-70.2904444444435 44.0642222222221,-70.2904444444435 44.064111111111,-70.2904444444435 44.0639999999999,-70.2905555555546 44.0638888888888,-70.2906666666657 44.0637777777777,-70.2907777777768 44.0636666666665,-70.2908888888879 44.0635555555554,-70.290999999999 44.0634444444443,-70.2911111111101 44.0633333333332,-70.2912222222212 44.0632222222221,-70.2913333333323 44.0632222222221,-70.2914444444435 44.0632222222221,-70.2915555555546 44.063111111111,-70.2916666666657 44.063111111111,-70.2917777777768 44.0629999999999,-70.2918888888879 44.0629999999999,-70.291999999999 44.0629999999999,-70.2921111111101 44.0629999999999,-70.2922222222212 44.063111111111,-70.2923333333324 44.063111111111,-70.2924444444435 44.063111111111,-70.2925555555546 44.063111111111,-70.2926666666657 44.0629999999999,-70.2927777777768 44.0629999999999,-70.2928888888879 44.0628888888888,-70.292999999999 44.0627777777777,-70.2931111111101 44.0626666666666,-70.2932222222212 44.0625555555554,-70.2933333333324 44.0624444444443,-70.2934444444435 44.0624444444443,-70.2935555555546 44.0624444444443,-70.2936666666657 44.0623333333332,-70.2937777777768 44.0622222222221,-70.2938888888879 44.062111111111,-70.293999999999 44.0619999999999,-70.2941111111101 44.0618888888888,-70.2942222222212 44.0617777777777,-70.2943333333323 44.0617777777777,-70.2944444444435 44.0617777777777,-70.2945555555546 44.0617777777777,-70.2946666666657 44.0618888888888,-70.2947777777768 44.0619999999999,-70.2948888888879 44.062111111111,-70.294999999999 44.0622222222221,-70.2951111111101 44.0622222222221)
```

The `strmOrder` column has an equivalent in NHD flowlines that is used for styling currently:

https://github.com/WikiWatershed/model-my-watershed/blob/878a4aa53845279a21dd2dcd8bdc8bacaf4d4ebf/src/tiler/styles/streams.mss#L17-L39

Make it so that:

- [x] The same blue color is used to show TDX streams as used for the NHD flowlines
- [ ] The same stream order rules are used for zoom dependent line width

**Comment by aufdenkampe** [comment]
@rajadain, thanks for creating this issue and highlighting the code with the rendering rules presently used for the NHD flow lines, based on [map zoom levels](https://wiki.openstreetmap.org/wiki/Zoom_levels).

We've done some thinking about this and are prepared to help, if you would like, but might need to exchange some info to be most effective. In particular, I'm thinking that we might want a few more levels of rendering than we use for NHD. Also, a TDX-hydro `strmOrder=1` is equivalent to NHD `stream_order=2`, etc., so we'll want to take that into account.

cc: @kieranbartels & @ptomasula

**Comment by rajadain** [comment]
I was able to use a script like this to bake static vector tiles that are restricted to certain stream orders: [ingest-vector-tiles-2.sh](https://github.com/user-attachments/files/15892475/ingest-vector-tiles-2.txt). You can see the cutoffs in the file, but they are:

| Zoom Levels | Stream Orders |
|-------------|---------------|
| 0-5         | 5+            |
| 6-8         | 4+            |
| 9-11        | 3+            |
| 12          | 2+            |
| 13          | 1+            |

This is what the results look like in MMW:

https://github.com/WikiWatershed/model-my-watershed/assets/1430060/f3639951-601d-4129-88e3-e99e5c7ba504

@aufdenkampe please let me know if I should use different cutoffs.

---

There is one intermediate zoom with a strange artifact that needs to be addressed, and I don't have zoom-specific stream order specific widths figured out yet, but it seems to be moving in the direction we want.

Also, I'm currently using uncompressed vector tiles, which for this subset is about 400 MB when made upto zoom level 13. We should look into compression as well to reduce hosting costs.

**Comment by rajadain** [comment]
@aufdenkampe @darscott the stream layer is available on staging here: https://staging.modelmywatershed.org/, if you scroll down the Streams tab in the layer selector and select "TDX Streamnet 7020038340" to see the full dataset, and "TDX Streamnet 7020038340 Ordered" to see the ones filtered by zoom level and stream order.

The current breaks are listed in [the comment above](https://github.com/WikiWatershed/model-my-watershed/issues/3625#issuecomment-2176969620). Please take a look and advise if these breaks are acceptable, or if we should make any tweaks.

Note that there are some odd artifacts if zoomed out too much. We're working on those, please ignore them for now.

**Comment by aufdenkampe** [comment]
@rajadain, I've just reviewed staging carefully and would like to make the following revision to the zoom rendering rules you shared in https://github.com/WikiWatershed/model-my-watershed/issues/3625#issuecomment-2176969620 and pasted below.

Currently on Staging:
| Zoom Levels | Stream Orders |
|-------------|---------------|
| 0-5         | 5+            |
| 6-8         | 4+            |
| 9-11        | 3+            |
| 12          | 2+            |
| 13          | 1+            |

My overall assessment is that streams are too dense at lower zoom levels (continent and country scales) and not dense enough at the highest zoom levels (local scales).

PROPOSED revision:
| Zoom Levels | Minimum Stream Order Rendered (x) | 
|---------|-------|
| 0-4        | 6         |
| 5-6        | 5         |
| 7-8        | 4         |
| 9-10      | 3         |
| 11           | 2         |
| 12-13     | 1         |

@rajadain, I'm sure you noticed that I expanded from 5 to 6 sets of tiles. If that is too much, then please group Zoom level 11 with 12-13.

I'll provide line width recommendations in the next comment, using Minimum Stream Order Rendered (x)  to create a formula.

**Comment by rajadain** [comment]
I made an [updated version of my script](https://github.com/user-attachments/files/16459189/ingest-vector-tiles-3.txt) that is [easier to parameterize](https://github.com/user-attachments/files/16459192/tdx-styling.csv), and ingested the Vector Tiles with the new breaks. These are headed to staging now, will post here when ready.

**Comment by rajadain** [comment]
I've also been playing around with line width. If we very naively just use the stream order for line width, the results are pretty inconsistent and hard to follow. I'm going to next investigate if this is a true artifact of the underlying data or if it's a result of our ingest process:


https://github.com/user-attachments/assets/12e24ccc-1eb9-475b-8bc4-083eb1258e24

**Comment by rajadain** [comment]
The underlying dataset is not the problem. Here's what it looks like when we draw by strmOrder from the source .gpkg in QGIS:

<img width="2672" alt="image" src="https://github.com/user-attachments/assets/fcf60720-1d92-4e3d-afb2-10f727f6a61f">

![Screenshot 2024-08-01 at 4 47 05 PM](https://github.com/user-attachments/assets/ef6ced11-48f7-425a-9e4c-0d9105245b74)

**Comment by rajadain** [comment]
I'm now ingesting this into Postgres to see what that looks like.

**Comment by rajadain** [comment]
I've ingested all of Africa and Alaska into Postgres locally, and the Windshaft performance is pretty good and it doesn't take as much as space as I'd thought it would: it's 8GB for the above, so think it would be around 25-30GB for the global dataset, which is the same order of magnitude as the high resolution NHD streams. This may be the way forward, and not the Vector Tiles ingested above.

Note that the below screenshots do not use the updated width styling that @aufdenkampe specified above, they're just reusing the styles for NHD. Will fix that next.

![Screen Shot 2024-08-05 at 16 40 37-fullpage](https://github.com/user-attachments/assets/a923e19f-14f1-43b9-8177-c17700ad257e)

![Screen Shot 2024-08-05 at 16 41 02-fullpage](https://github.com/user-attachments/assets/1fdf1b5f-b601-4165-8a89-cd8c368c1837)

![Screen Shot 2024-08-05 at 16 41 21-fullpage](https://github.com/user-attachments/assets/9260daf0-b0d0-4f17-acbf-4488b748a929)

**Comment by aufdenkampe** [comment]
@rajadain, these look great! Thanks for moving it all forward. I won't be able to provide much additional guidance, so please use your judgement to make the decisions required to get this done before you head out this week. Thanks!!!!

**Comment by aufdenkampe** [comment]
@rajadain, since we'll be re-ingesting all the stream data next week, I want to reopen this issue to further explore stream line width rendering. As you found in your https://github.com/WikiWatershed/model-my-watershed/issues/3625#issuecomment-2263444013 above, the function that relates stream order to line width needs to be zoom dependent too.

Can you point me to the code you're currently using to set line widths as you create the tiles?

**Comment by rajadain** [comment]
Here's the current TDX Hydro Styling code:

https://github.com/WikiWatershed/model-my-watershed/blob/972905c4bc44c6adfdb87b77592fa44c16e724ac/src/tiler/styles/streams.mss#L280-L362

**Comment by aufdenkampe** [comment]
Thanks! What does `@zoomBase` refer to? A single value? Or is it the current zoom level of the browser?

**Comment by rajadain** [comment]
Apparently it's a single value set to `0.6`:

https://github.com/WikiWatershed/model-my-watershed/blob/972905c4bc44c6adfdb87b77592fa44c16e724ac/src/tiler/styles/_variables.mss#L1

Likely so we can adjust all styles at once by tweaking this one number. However, since changing it would impact styles from other layers too, I think we'll likely keep it at `0.6`.

**Comment by aufdenkampe** [comment]
@rajadain, based on some experimentation, I'm exploring line width rendering something like this table, which aligns with my zoom-rendering recommendations in https://github.com/WikiWatershed/model-my-watershed/issues/3625#issuecomment-2260591157

The column headings 1-8 are Stream Orders, with line widths in the cells below.

| Zoom Levels | Minimum Stream Order Rendered (x) |  1  |  2  |  3  |  4  |  5  |  6  | 7 | 8 |
|-------------|-----------------------------------|:---:|:---:|:---:|:---:|:---:|:---:|:-:|:-:|
| 0-4         | 6                                 | n/a | n/a | n/a | n/a | n/a | 0.5 | 1 | 2 |
| 5-6         | 5                                 | n/a | n/a | n/a | n/a | 0.5 |  1  | 2 | 3 |
| 7-8         | 4                                 | n/a | n/a | n/a | 0.5 |  1  |  2  | 3 | 4 |
| 9-10        | 3                                 | n/a | n/a | 0.5 |  1  |  2  |  3  | 4 | 5 |
| 11          | 2                                 | n/a | 0.5 |  1  |  2  |  3  |  4  | 5 | 6 |
| 12-13       | 1                                 | 0.5 |  1  |  2  |  3  |  4  |  5  | 6 | 7 |

**Comment by rajadain** [comment]
@aufdenkampe the above recommended styling is now applied to the TDX Hydro dataset on staging: https://staging.modelmywatershed.org/. The code for this is in https://github.com/WikiWatershed/model-my-watershed/commit/d6977332ca3e1d3e4031a99bd608dae1a1c6b81f. Let me know what you think. I think `0.5` is too fine, we should start with `1` being the minimum. There's also some places where certain streams appear darker on lower zoom levels, but lighter on higher zoom levels, which I don't understand. But wanted you to take a look.

**Comment by rajadain** [comment]
Here's a demo video:

https://github.com/user-attachments/assets/6d1d9835-56ec-441d-92d6-92e07aa1823e

**Comment by aufdenkampe** [comment]
@rajadain, thanks for getting all this onto staging and creating a quick video.  I just created a bunch of screenshots, so I could better explore each zoom level, and saved them here: [TDX_stream_styling](https://drive.google.com/drive/folders/1r4XE94Afs-vTO4zQ7Nfb3nxmlMxIF58v?usp=drive_link).

Here are a few initial thoughts.
- Zooms 0-4 don't show any streams, and the "TDX Hydro" layer selector option is greyed out. Let's do show streams for the entire globe, at the minimum for marketing!
- Zoom levels 5, 7 & 12 look perfect.
- Zoom levels 11, 13, 14 seem to have bugs in rendering, that I think you noticed.
- Other zoom levels are acceptable, but generally seem to be thinner than I would expect, even for middle-range streams.

I see in your commit d697733 that you multiply all line-widths by `@zoomBase` , which equals 0.6. Is that necessary? If so, could that be causing the odd behaviors and thinner than expected lines? 

Do we need to divide all my numbers in https://github.com/WikiWatershed/model-my-watershed/issues/3625#issuecomment-2371838760 by 0.6 to compensate for `* @zoomBase`? I see that appears to be more or less how the current rendering works:

https://github.com/WikiWatershed/model-my-watershed/blob/953fc155772f4f9c93a14d10af6a41275ede245e/src/tiler/styles/streams.mss#L56-L62
